backend/review/annotations.py

```diff
@@ -0,0 +1,116 @@
+from django.db.models import Avg, Count, FloatField, Max, Subquery, Value
+
+from review.models import ALL_FIELD_SLUGS, Review, ReviewBit
+
+
+"""
+Queryset annotations
+====================
+
+Actual review data is stored in an Entity-Attribute-Value (EAV) format in the ReviewBit
+model. This means that getting associated review data for a queryset requires a few
+JOINs under the hood. Doing aggregations on these ReviewBits also requires the explicit
+use of subqueries. You can read about Subqueries here:
+https://docs.djangoproject.com/en/2.2/ref/models/expressions/#subquery-expressions.
+
+In short, however, subqueries allow us to query for review data from the ReviewBit table
+*inside* any other queryset to use in aggregations and annotations. We can filter down
+the ReviewBits that we want to aggregate based on their field name, along with any other
+Django filter query that can be different *per row* in the outer query. To match on fields
+from the outer query, we use the OuterRef() expressions.
+
+This allows us to have the database do all of the work of averaging PCR data. Were we to do
+this aggregation all in Python code, it would likely take many more queries (read: round-trips to
+the DB), be *much* slower, and require cacheing.
+"""
+
+
+def review_averages(queryset, subfilters, fields=None, prefix="", semester_aggregations=False):
+    """
+    Annotate the queryset with the average of all ReviewBits matching the given subfilters.
+    :param queryset: Queryset to annotate with averages.
+    :param subfilters: Filters to filter down the ReviewBits used in each individual aggregation.
+    use OuterRef() to refer to values in the outer queryset.
+    :param fields: the ReviewBit fields to aggregate. if None, defaults to the four fields
+    used in PCP.
+    :param prefix: prefix for fields in annotated queryset. Useful when applying review_averages
+    multiple times to the same queryset with different subfilters.
+    :param semester_aggregations: option to annotate additional semester aggregations for the
+    semester returned by the subfilters (only useful if subfilters filter down to one semester),
+    as well as the count of the number of semesters included in the queryset's annotations.
+    """
+    if fields is None:
+        fields = ["course_quality", "difficulty", "instructor_quality", "work_required"]
+    queryset = queryset.annotate(
+        **{
+            (prefix + field): Subquery(
+                ReviewBit.objects.filter(field=field, **subfilters)
+                .values("field")
+                .order_by()
+                .annotate(avg=Avg("average"))
```

Comment:
I know this is old code but I'm wondering where the ReviewBit score field is referenced?  Also, maybe I missed this in the Django docs, but where is the field `average` coming from for `Avg("average")` to work?  Conceptually I get why you would want `Avg("average")` in this line to return the average score for the field group but I'm not sure why it would with this code...

Reply:
In the last PR, the `score` field got renamed to `average`. So `Avg("average")` is annotating the average of all the scores from all the filtered ReviewBits. The reason the field was renamed was because in the big PCR dump we have a lot more information on Reviews, including median, standard deviation, and distribution that we can all load in.

Reply:
That's awesome.  Great to hear we have that other data.

Reply:
Wait what would distribution be?

Reply:
We can see how many people voted 0, 1, 2, 3 and 4 for each field